filters/CovarianceFeaturesFilter.cpp

```diff
@@ -65,28 +66,179 @@ std::string CovarianceFeaturesFilter::getName() const
+                "Verticality", Dimension::Type::Double);
+        if (m_featureTypes & FeatureType::Omnivariance)
+            m_extraDims["Omnivariance"] = layout->registerOrAssignDim(
+                "Omnivariance", Dimension::Type::Double);
+        if (m_featureTypes & FeatureType::Anisotropy)
+            m_extraDims["Anisotropy"] = layout->registerOrAssignDim(
+                "Anisotropy", Dimension::Type::Double);
+        if (m_featureTypes & FeatureType::Eigenentropy)
+            m_extraDims["Eigenentropy"] = layout->registerOrAssignDim(
+                "Eigenentropy", Dimension::Type::Double);
+        if (m_featureTypes & FeatureType::Sum)
+            m_extraDims["Sum"] =
+                layout->registerOrAssignDim("Sum", Dimension::Type::Double);
+        if (m_featureTypes & FeatureType::SurfaceVariation)
+            m_extraDims["SurfaceVariation"] = layout->registerOrAssignDim(
+                "SurfaceVariation", Dimension::Type::Double);
+        if (m_featureTypes & FeatureType::DemantkeVerticality)
+            m_extraDims["DemantkeVerticality"] = layout->registerOrAssignDim(
+                "DemantkeVerticality", Dimension::Type::Double);
+        if (m_featureTypes & FeatureType::Density)
+            m_extraDims["Density"] =
+                layout->registerOrAssignDim("Density", Dimension::Type::Double);
     }
 }
 
-void CovarianceFeaturesFilter::filter(PointView& view)
+void CovarianceFeaturesFilter::initialize()
 {
+    for (auto& feat : m_features)
+    {
+        std::string f = Utils::tolower(feat);
+        Utils::trim(f);
+        if (f == "all")
+        {
+            m_featureTypes = ~0;
+            break;
+        }
+        if (f == "linearity")
+            m_featureTypes |= FeatureType::Linearity;
+        else if (f == "planarity")
+            m_featureTypes |= FeatureType::Planarity;
+        else if (f == "scattering")
+            m_featureTypes |= FeatureType::Scattering;
+        else if (f == "verticality")
+            m_featureTypes |= FeatureType::Verticality;
+        else if (f == "omnivariane")
+            m_featureTypes |= FeatureType::Omnivariance;
+        else if (f == "anisotropy")
+            m_featureTypes |= FeatureType::Anisotropy;
+        else if (f == "eigenentropy")
+            m_featureTypes |= FeatureType::Eigenentropy;
+        else if (f == "sum")
+            m_featureTypes |= FeatureType::Sum;
+        else if (f == "surfacevariation")
+            m_featureTypes |= FeatureType::SurfaceVariation;
+        else if (f == "demantkeverticality")
+            m_featureTypes |= FeatureType::DemantkeVerticality;
+        else if (f == "density")
+            m_featureTypes |= FeatureType::Density;
+        else
+            throwError("Invalid feature type: '" + f + "'.");
+    }
+}
 
+void CovarianceFeaturesFilter::prepared(PointTableRef table)
+{
+    const PointLayoutPtr layout(table.layout());
+    if (m_optimal)
```

Comment:
Where? The stage docs? It's there. Is the wording unclear?

Reply:
The only way this is going to work is if you precede this with the OptimalNeighborhoodFilter. I think I would just say that. And that it will fail if you don't have that in your pipeline.

Reply:
Does it address your concern for the description of `optimized` at https://github.com/PDAL/PDAL/pull/3180/files#diff-e73bc2bdaf15a0653aebb7bf7fe00c4dR113-R116 to be rewritten as:

```rst
optimized
    ``optimized`` can be set to ``true`` to enable computation of features using precomputed
    optimal neighborhoods. Requires :ref:`filters.optimalneighborhood` be run prior to this
    stage. Enables computation of ``Density`` feature. [Default: false]
```

Reply:
Fine.

Reply:
Fixed via https://github.com/PDAL/PDAL/pull/3180/commits/181546e9ce43e88caa401aaa70c4680bccf5f33c